### PR TITLE
Remove typed_pattern

### PR DESCRIPTION
**Pull request by mattn**
See https://github.com/prabirshrestha/vim-lsp/pull/701

I realized typed_pattern is not good to remove prefix of items in completion. To fix correctly, we should implement same as asyncomplete. If we set b:asyncomplete_refresh_pattern correctly, this problem should be fixed.

However, I don't think this PR must not be merged as is. `b:asyncomplete_refresh_pattern` should be more general name for that another plugins can be used.

What do you think?

### CODE REVIEW

**Comment by thomasfaingnaert** [comment]
Will this still work if one uses omnicomplete instead of asyncomplete?

**Comment by mattn** [comment]
Yes, I checked this without asyncomplete.vim.

**Comment by prabirshrestha** [comment]
Should we name it `vim_lsp_refresh_pattern`?

I don't think other plugins need to care about this. In asyncomplete I actually detect if the popup was created by asyncomplete or by other mode such as omni. https://github.com/prabirshrestha/asyncomplete.vim/blob/52f857a4c4c51abf6b29f4ce0b3db0aae992f619/autoload/asyncomplete.vim#L438-L442. If I detect other mode I skip processing keys in asyncomplete.

**Comment by mattn** [comment]
asyncomplete can be fixed about this like below, I think.

```diff
diff --git a/autoload/asyncomplete.vim b/autoload/asyncomplete.vim
index 373f612..5361ff8 100644
--- a/autoload/asyncomplete.vim
+++ b/autoload/asyncomplete.vim
@@ -478,6 +478,14 @@ function! asyncomplete#preprocess_complete(ctx, items)
     endif
     call asyncomplete#log('core', 'asyncomplete#preprocess_complete calling complete()', a:ctx['startcol'], a:items)
+
+    let l:prefix = getline('.')[a:ctx['startcol']:col('.')]
+    for i in range(len(a:items))
+      let l:item = a:items[i]
+      if stridx(l:item.word, l:prefix) == 0
+        let l:item.word = l:item.word[l:prefix :]
+      endif
+    endfor
     call complete(a:ctx['startcol'], a:items)
 endfunction
```

File: autoload/lsp/ui/vim/completion.vim [hunk 1]
```diff
@@ -191,10 +191,7 @@ function! s:clear_inserted_text(line, position, completed_item, completion_item)
 
   " Expand remove range to textEdit.
   if has_key(a:completion_item, 'textEdit')
-    let l:range['start']['character'] = min([
-          \   l:range['start']['character'],
-          \   a:completion_item['textEdit']['range']['start']['character']
-          \ ])
+    let l:range['start']['character'] = a:completion_item['textEdit']['range']['start']['character']
```
Comment (hrsh7th) [inline autoload/lsp/ui/vim/completion.vim]: I discussed with mattn. It is OK.
I checked with `vim-language-server`, `intelephense`, `html-languageserver` works all.

Comment (mattn) [reply autoload/lsp/ui/vim/completion.vim]: Thank you! I checked metals, julia-language-server, css-languageserver, clangd.

**Comment by mattn** [comment]
Totally I'm okay to merge this. Current implementation use startcol to remove pre-typed text. And the startcol is calculated with regexp pattern `vim_lsp_refresh_pattern`. @prabirshrestha Do you prefer the variable is buffer variable? or configuration `"config": {"refresh_pattern": "xxx"}` ?

**Comment by mattn** [comment]
Currently, I found some filetypes to specify refresh_pattern to fix completion.

## css
```
      "refresh_pattern": "\\([a-zA-Z0-9_-]\\+\\)$",
```
Because property-name contains `-` (ex: "background-color") but syntax/css.vim does not add `-` into iskeyword.

## html
```
      "refresh_pattern": "\\(/\\|k\\+\\)$",
```
Because close tag `</` makes dupplicated slash `<//div>`. Most of users does not want to add `/` to iskeyword.

## json
```
      "refresh_pattern": "\\(\"\\k*\\|\\[\\|\\k\\+\\)$",
```
Because `[` should trigger completion for array. This does not be wanted from users to add to iskeyword.

## ps1
```
      "refresh_pattern": "\\(\\$[a-zA-Z0-9_:]*\\|\\k\\+\\)$",
```
Because `$` should trigger completion for `$ENV`. ditto.

I already provide those patterns from vim-lsp-settings.

**Comment by clason** [comment]
I tested texlab (which is always a bit of a stress test; LaTeX being quite different from usual programming languages), and it seems to work fine as well.

**Comment by prabirshrestha** [comment]
@mattn config variable is better. Also need to be careful about multiple langservers for same buffer.

**Comment by mattn** [comment]
Added code that use refresh_pattern which found a first item from server configs.

**Comment by mattn** [comment]
Thanks all!

**Pull request merged**